File: guide-api-agent-context/agent_run_keypair_jwt.py
#!/usr/bin/env python3
"""
Key-Pair JWT authentication for the Snowflake agent:run API.

Generates RS256-signed JWTs from an RSA private key, suitable for service
accounts, CI/CD pipelines, or any environment where password/PAT auth is
not available.

Requirements:
    pip install cryptography requests

Environment Variables:
    SNOWFLAKE_ACCOUNT:          Account identifier (e.g., 'myorg-myaccount')
    SNOWFLAKE_USER:             Snowflake username with RSA public key assigned
    SNOWFLAKE_PRIVATE_KEY_PATH: Path to PEM-encoded RSA private key file

Snowflake setup (one-time):
    -- 1. Generate key pair
    openssl genrsa -out rsa_key.pem 2048
    openssl rsa -in rsa_key.pem -pubout -out rsa_key.pub

    -- 2. Assign public key to user (run as ACCOUNTADMIN)
    ALTER USER my_user SET RSA_PUBLIC_KEY='<contents of rsa_key.pub without header/footer>';

See also:
    https://docs.snowflake.com/en/user-guide/key-pair-auth
    https://docs.snowflake.com/en/developer-guide/sql-api/authenticating
"""

import hashlib
import json
import math
import os
import sys
import time
from typing import Optional, Tuple

import requests
from cryptography.hazmat.primitives import hashes, serialization
from cryptography.hazmat.primitives.asymmetric import padding


# ---------------------------------------------------------------------------
# JWT generation
# ---------------------------------------------------------------------------

def _base64url(data: bytes) -> str:
    import base64
    return base64.urlsafe_b64encode(data).rstrip(b"=").decode("ascii")


def _normalize_account(account: str) -> str:
    """ACCOUNT-USER format required by Snowflake JWT: uppercase, dots→hyphens."""
    return (
        account.strip()
        .replace(".snowflakecomputing.com", "")
        .replace(".", "-")
        .upper()
    )


def _build_fingerprint(private_key_pem: bytes) -> str:
    """SHA256 fingerprint of the SPKI-encoded public key (base64)."""
    private_key = serialization.load_pem_private_key(private_key_pem, password=None)
    public_der = private_key.public_key().public_bytes(
        serialization.Encoding.DER,
        serialization.PublicFormat.SubjectPublicKeyInfo,
    )
    import base64
    return base64.b64encode(hashlib.sha256(public_der).digest()).decode("ascii")


def generate_keypair_jwt(
    account: str,
    user: str,
    private_key_pem: bytes,
    expires_in_seconds: int = 3600,
) -> Tuple[str, int]:
    """
    Build and sign a Snowflake KEYPAIR_JWT token.

    Returns (token_string, expiry_unix_timestamp).
    """
    now = math.floor(time.time())
    exp = now + expires_in_seconds

    normalized_account = _normalize_account(account)
    username = user.strip().upper()
    qualified = f"{normalized_account}.{username}"
    fingerprint = _build_fingerprint(private_key_pem)

    header = {"alg": "RS256", "typ": "JWT"}
    payload = {
        "iss": f"{qualified}.SHA256:{fingerprint}",
        "sub": qualified,
        "iat": now,
        "exp": exp,
    }

    signing_input = f"{_base64url(json.dumps(header).encode())}.{_base64url(json.dumps(payload).encode())}"

    private_key = serialization.load_pem_private_key(private_key_pem, password=None)
    signature = private_key.sign(
        signing_input.encode(),
        padding.PKCS1v15(),
        hashes.SHA256(),
    )

    token = f"{signing_input}.{_base64url(signature)}"
    return token, exp


# ---------------------------------------------------------------------------
# Token cache -- reuse until 5 minutes before expiry
# ---------------------------------------------------------------------------

_cached_token: Optional[str] = None
_cached_exp: int = 0


def get_jwt(account: str, user: str, private_key_pem: bytes) -> str:
    """Return a cached JWT, refreshing when within 5 minutes of expiry."""
    global _cached_token, _cached_exp
    now = math.floor(time.time())
    if _cached_token and (_cached_exp - 300) > now:
        return _cached_token
    _cached_token, _cached_exp = generate_keypair_jwt(account, user, private_key_pem)
    return _cached_token


# ---------------------------------------------------------------------------
# Snowflake helpers
# ---------------------------------------------------------------------------

def _build_headers(jwt: str) -> dict:
    return {
        "Authorization": f"Bearer {jwt}",
        "X-Snowflake-Authorization-Token-Type": "KEYPAIR_JWT",
        "Content-Type": "application/json",
    }


def create_thread(account: str, jwt: str) -> str:
    url = f"https://{account.strip()}.snowflakecomputing.com/api/v2/cortex/threads"
    resp = requests.post(url, headers=_build_headers(jwt), json={"origin_application": "keypair_jwt_example"})
    resp.raise_for_status()
    return resp.json()["id"]


def run_agent(
    account: str,
    jwt: str,
    database: str,
    schema: str,
    agent_name: str,
    thread_id: str,
    message: str,
    role: Optional[str] = None,
    warehouse: Optional[str] = None,
) -> None:
    """Call agent:run with key-pair JWT auth and stream the response."""
    url = (
        f"https://{account.strip()}.snowflakecomputing.com"
        f"/api/v2/databases/{database}/schemas/{schema}/agents/{agent_name}:run"
    )

    headers = _build_headers(jwt)
    if role:
        headers["X-Snowflake-Role"] = role
    if warehouse:
        headers["X-Snowflake-Warehouse"] = warehouse

    payload = {
        "thread_id": thread_id,
        "parent_message_id": 0,
        "messages": [{"role": "user", "content": [{"type": "text", "text": message}]}],
    }

    print(f"\n{'='*70}")
    print(f"Agent:  {database}.{schema}.{agent_name}")
    print(f"Auth:   Key-Pair JWT")
    if role:
        print(f"Role:   {role}")
    if warehouse:
        print(f"WH:     {warehouse}")
    print(f"Q:      {message}")
    print(f"{'='*70}\n")

    event_type = ""
    with requests.post(url, headers=headers, json=payload, stream=True) as resp:
        resp.raise_for_status()
        for line in resp.iter_lines():
            if not line:
                continue
            decoded = line.decode("utf-8")
            if decoded.startswith("event:"):
                event_type = decoded.split(":", 1)[1].strip()
            elif decoded.startswith("data:"):
                try:
                    data = json.loads(decoded.split(":", 1)[1].strip())
                    if event_type == "response.text.delta":
                        print(data.get("text", ""), end="", flush=True)
                    elif event_type == "error":
                        print(f"\n[ERROR] {data.get('message', 'Unknown')}")
                except json.JSONDecodeError:
                    pass
    print()


# ---------------------------------------------------------------------------
# Main
# ---------------------------------------------------------------------------

def main():
    account = os.getenv("SNOWFLAKE_ACCOUNT")
    user = os.getenv("SNOWFLAKE_USER")
    key_path = os.getenv("SNOWFLAKE_PRIVATE_KEY_PATH")

    if not account or not user or not key_path:
        print("Required environment variables:")
        print("  SNOWFLAKE_ACCOUNT          e.g. myorg-myaccount")
        print("  SNOWFLAKE_USER             e.g. MY_SERVICE_USER")
        print("  SNOWFLAKE_PRIVATE_KEY_PATH e.g. ./rsa_key.pem")
        sys.exit(1)

    private_key_pem = open(key_path, "rb").read()

    print("Generating key-pair JWT...")
    jwt = get_jwt(account, user, private_key_pem)
    print("OK  JWT generated")

    print("Creating thread...")
    thread_id = create_thread(account, jwt)
    print(f"OK  Thread: {thread_id}")

    # --- Customize these to match your agent ---
    run_agent(
        account=account,
        jwt=jwt,
        database="MYDB",
        schema="MYSCHEMA",
        agent_name="my_agent",
        thread_id=thread_id,
        message="What were the top 5 products by revenue last month?",
        role="ANALYST_ROLE",
        warehouse="COMPUTE_WH",
    )


if __name__ == "__main__":
    main()

File: guide-api-agent-context/agent_run_keypair_jwt.js
#!/usr/bin/env node
/**
 * Key-Pair JWT authentication for the Snowflake agent:run API.
 *
 * Zero external dependencies -- uses only Node.js built-in `crypto`.
 * Suitable for service accounts, CI/CD, or any environment where
 * password/PAT auth is not available.
 *
 * Environment Variables:
 *   SNOWFLAKE_ACCOUNT          Account identifier (e.g. 'myorg-myaccount')
 *   SNOWFLAKE_USER             Snowflake username with RSA public key assigned
 *   SNOWFLAKE_PRIVATE_KEY_PATH Path to PEM-encoded RSA private key
 *
 * Snowflake setup (one-time):
 *   openssl genrsa -out rsa_key.pem 2048
 *   openssl rsa -in rsa_key.pem -pubout -out rsa_key.pub
 *   ALTER USER my_user SET RSA_PUBLIC_KEY='<pub key without header/footer>';
 *
 * See also:
 *   https://docs.snowflake.com/en/user-guide/key-pair-auth
 *   https://docs.snowflake.com/en/developer-guide/sql-api/authenticating
 */

const fs = require('fs');
const { createHash, createPublicKey, createSign } = require('crypto');

// ---------------------------------------------------------------------------
// JWT generation (no dependencies)
// ---------------------------------------------------------------------------

function normalizeAccount(account) {
  return account
    .trim()
    .replace('.snowflakecomputing.com', '')
    .replace(/\./g, '-')
    .toUpperCase();
}

function buildFingerprint(privateKeyPem) {
  const publicKeyDer = createPublicKey(privateKeyPem).export({
    type: 'spki',
    format: 'der',
  });
  return createHash('sha256').update(publicKeyDer).digest('base64');
}

function toBase64Url(input) {
  return Buffer.from(input).toString('base64url');
}

function signJwt({ account, user, privateKeyPem, expiresInSeconds = 3600 }) {
  const now = Math.floor(Date.now() / 1000);
  const exp = now + expiresInSeconds;

  const normalizedAccount = normalizeAccount(account);
  const username = user.trim().toUpperCase();
  const qualified = `${normalizedAccount}.${username}`;

  const normalizedKey = privateKeyPem.replace(/\\n/g, '\n').trim();
  const fingerprint = buildFingerprint(normalizedKey);

  const header = { alg: 'RS256', typ: 'JWT' };
  const payload = {
    iss: `${qualified}.SHA256:${fingerprint}`,
    sub: qualified,
    iat: now,
    exp,
  };

  const signingInput = `${toBase64Url(JSON.stringify(header))}.${toBase64Url(JSON.stringify(payload))}`;
  const signature = createSign('RSA-SHA256')
    .update(signingInput)
    .end()
    .sign(normalizedKey, 'base64url');

  return { token: `${signingInput}.${signature}`, exp };
}

// Token cache -- reuse until 5 minutes before expiry
let cachedJwt = null;

function getJwt(account, user, privateKeyPem) {
  const now = Math.floor(Date.now() / 1000);
  if (cachedJwt && cachedJwt.exp - 300 > now) {
    return cachedJwt.token;
  }
  cachedJwt = signJwt({ account, user, privateKeyPem });
  return cachedJwt.token;
}

// ---------------------------------------------------------------------------
// Snowflake helpers
// ---------------------------------------------------------------------------

function buildHeaders(jwt) {
  return {
    Authorization: `Bearer ${jwt}`,
    'X-Snowflake-Authorization-Token-Type': 'KEYPAIR_JWT',
    'Content-Type': 'application/json',
  };
}

async function createThread(account, jwt) {
  const url = `https://${account.trim()}.snowflakecomputing.com/api/v2/cortex/threads`;
  const resp = await fetch(url, {
    method: 'POST',
    headers: buildHeaders(jwt),
    body: JSON.stringify({ origin_application: 'keypair_jwt_example' }),
  });
  if (!resp.ok) throw new Error(`Thread creation failed: ${resp.status} ${await resp.text()}`);
  const data = await resp.json();
  return data.id ?? data.thread_id;
}

async function runAgent({
  account, jwt, database, schema, agentName,
  threadId, message, role, warehouse,
}) {
  const base = `https://${account.trim()}.snowflakecomputing.com`;
  const url = `${base}/api/v2/databases/${database}/schemas/${schema}/agents/${agentName}:run`;

  const headers = buildHeaders(jwt);
  if (role) headers['X-Snowflake-Role'] = role;
  if (warehouse) headers['X-Snowflake-Warehouse'] = warehouse;

  const body = {
    thread_id: threadId,
    parent_message_id: 0,
    messages: [{ role: 'user', content: [{ type: 'text', text: message }] }],
  };

  console.log(`\n${'='.repeat(70)}`);
  console.log(`Agent:  ${database}.${schema}.${agentName}`);
  console.log(`Auth:   Key-Pair JWT`);
  if (role) console.log(`Role:   ${role}`);
  if (warehouse) console.log(`WH:     ${warehouse}`);
  console.log(`Q:      ${message}`);
  console.log(`${'='.repeat(70)}\n`);

  const resp = await fetch(url, {
    method: 'POST',
    headers,
    body: JSON.stringify(body),
  });
  if (!resp.ok) throw new Error(`agent:run failed: ${resp.status} ${await resp.text()}`);

  const reader = resp.body.getReader();
  const decoder = new TextDecoder();
  let eventType = '';

  while (true) {
    const { done, value } = await reader.read();
    if (done) break;
    const chunk = decoder.decode(value);
    for (const line of chunk.split('\n')) {
      if (line.startsWith('event:')) {
        eventType = line.split(':')[1].trim();
      } else if (line.startsWith('data:')) {
        try {
          const data = JSON.parse(line.slice(5).trim());
          if (eventType === 'response.text.delta') {
            process.stdout.write(data.text || '');
          } else if (eventType === 'error') {
            console.error(`\n[ERROR] ${data.message || 'Unknown'}`);
          }
        } catch { /* skip non-JSON lines */ }
      }
    }
  }
  console.log();
}

// ---------------------------------------------------------------------------
// Exports (for use as a module in other projects)
// ---------------------------------------------------------------------------

module.exports = {
  normalizeAccount,
  buildFingerprint,
  signJwt,
  getJwt,
  buildHeaders,
};

// ---------------------------------------------------------------------------
// Main (when run directly)
// ---------------------------------------------------------------------------

async function main() {
  const account = process.env.SNOWFLAKE_ACCOUNT;
  const user = process.env.SNOWFLAKE_USER;
  const keyPath = process.env.SNOWFLAKE_PRIVATE_KEY_PATH;

  if (!account || !user || !keyPath) {
    console.error('Required environment variables:');
    console.error('  SNOWFLAKE_ACCOUNT          e.g. myorg-myaccount');
    console.error('  SNOWFLAKE_USER             e.g. MY_SERVICE_USER');
    console.error('  SNOWFLAKE_PRIVATE_KEY_PATH e.g. ./rsa_key.pem');
    process.exit(1);
  }

  const privateKeyPem = fs.readFileSync(keyPath, 'utf8');

  console.log('Generating key-pair JWT...');
  const jwt = getJwt(account, user, privateKeyPem);
  console.log('OK  JWT generated');

  console.log('Creating thread...');
  const threadId = await createThread(account, jwt);
  console.log(`OK  Thread: ${threadId}`);

  // --- Customize these to match your agent ---
  await runAgent({
    account,
    jwt,
    database: 'MYDB',
    schema: 'MYSCHEMA',
    agentName: 'my_agent',
    threadId,
    message: 'What were the top 5 products by revenue last month?',
    role: 'ANALYST_ROLE',
    warehouse: 'COMPUTE_WH',
  });
}

if (require.main === module) {
  main().catch((err) => {
    console.error(err);
    process.exit(1);
  });
}
